Improve usage of `Type.GetType` when activating types in data protection

khellang opened the pull request:
Fixes #54253
Fixes #48910

There are some documented cases where `Type.GetType(<name>, throwOnError: false)` can still throw:

>  The `throwOnError` parameter specifies what happens when the type is not found, and also suppresses certain other exception conditions, as described in the Exceptions section. Some exceptions are thrown regardless of the value of `throwOnError`. For example, if the type is found but cannot be loaded, a `System.TypeLoadException` is thrown even if `throwOnError` is `false`.

This happens in the case where a `decryptorType` of previous version of an assembly has been used and prevents a custom `IActivator` implementation from doing proper forwarding.

This PR handles this case by catching errors from `Type.GetType` and letting the activator handle it instead.

Another way of solving this could be to make `TypeForwardingActivator` `public` and expose a way to register other namespaces to forward. Today, the [`DecryptorTypeForwardingActivator`](https://github.com/Azure/azure-sdk-for-net/blob/55d99db1548b45c52f63c89ac1671a00bf486842/sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/DecryptorTypeForwardingActivator.cs) type in `Azure.Extensions.AspNetCore.DataProtection.Keys` is more or less an outdated copy of the `TypeForwardingActivator` in this repo.

// @JamesNK @amcasey

khellang committed:
Added test demonstrating behavior on TypeLoadException

khellang committed:
Catch exceptions when calling GetType and fall back to IActivator.CreateInstance<IXmlDecryptor>

khellang committed:
Move trimming suppression

amcasey commented on src/DataProtection/DataProtection/src/XmlEncryption/XmlEncryptionExtensions.cs:
```diff
@@ -95,6 +97,26 @@ private static IXmlDecryptor CreateDecryptor(IActivator activator, string decryp
+            : decryptorTypeName;
+        try
+        {
+            type = _getType(resolvedTypeName);
```
Nit: I like the comment in your PR description about how this can still throw when `throwIfFalse` is set.  Can you please put something like that here?

khellang replied:
Done!

amcasey reviewed (approved):
LGTM, but let's give @JamesNK time to comment.  Thanks!

khellang committed:
Include comment about how throwOnError: false can still throw

JamesNK commented on src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.Tests/XmlEncryption/XmlEncryptionExtensionsTests.cs:
```diff
@@ -49,6 +49,37 @@ public void DecryptElement_RootNodeRequiresDecryption_Success()
+        // We want to simulate an error loading the specified decryptor type, i.e.
+        // Could not load file or assembly 'Azure.Extensions.AspNetCore.DataProtection.Keys,
+        // Version=1.2.2.0, Culture=neutral, PublicKeyToken=92742159e12e44c8' or one of its dependencies.
+        XmlEncryptionExtensions._getType = _ => throw new TypeLoadException();
```
I worry about static state like this impacting other unit tests. It's not being reset so unit tests run afterwards would run this func. And if it was reset, there is still the possibility of concurrent tests.

Is it necessary? Can a value be specified that would trigger TypeLoadException without having to override how the GetType call worked.

khellang replied:
I did try to find a way to produce a `FileLoadException` or `TypeLoadException` by modifying the type, namespace, assembly, version etc. of the `decryptorType`, but wasn't able to. We're probably seeing this more because we're using .NET Framework with (semi-)manual binding redirects, while .NET (Core) seems much more forgiving.

If you have any suggestions, I'm all ears 😅

JamesNK replied:
.NET Core probably changed the behavior so that `throwOnError` was always respected.

Not having a unit test is one solution. However, when I wrote this code I didn't consider that an error could be thrown here. Catching it is a band aid.

Maybe the solution here is to match to these known types from a string without using `Type.GetType`. If the type name and assembly match, then we go directly to the known types. We'd be getting into the type name parsing business though, which is scary.

khellang replied:
> .NET Core probably changed the behavior so that `throwOnError` was always respected.

It's still documented to throw under the same circumstances, but .NET (Core) is much more resilient when it comes to resolving multiple versions of the same assemblies.

> Maybe the solution here is to match to these known types from a string without using Type.GetType. If the type name and assembly match, then we go directly to the known types. We'd be getting into the type name parsing business though, which is scary.

Yeah, why were the type checks introduced in the first place? I see there's a similar issue in `XmlKeyManager.CreateDeserializer`.

I have a crazy idea; what about an internal (or maybe even public?) interface

```csharp
internal interface ITypeForwardingActivator : IActivator
{
    bool TryForwardType(string decryptorTypeName, [NotNullWhen(true)] out Type? type);
}
```

which could be checked by pattern matching

```csharp
private static IXmlDecryptor CreateDecryptor(IActivator activator, string decryptorTypeName)
{
    if (activator is ITypeForwardingActivator typeForwarder && typeForwarder.TryForwardType(decryptorTypeName, out var type))
    {
        if (type == typeof(DpapiNGXmlDecryptor))
        {
            return activator.CreateInstance<DpapiNGXmlDecryptor>(decryptorTypeName);
        }
        else if (type == typeof(DpapiXmlDecryptor))
        {
            return activator.CreateInstance<DpapiXmlDecryptor>(decryptorTypeName);
        }
        else if (type == typeof(EncryptedXmlDecryptor))
        {
            return activator.CreateInstance<EncryptedXmlDecryptor>(decryptorTypeName);
        }
        else if (type == typeof(NullXmlDecryptor))
        {
            return activator.CreateInstance<NullXmlDecryptor>(decryptorTypeName);
        }
    }

    return activator.CreateInstance<IXmlDecryptor>(decryptorTypeName);
}
```

This would be implemented by the default `TypeForwardingActivator` and reuse the same existing logic.

Then the unit test could provide its own faked/mocked activator that returns `false` from `TryForwardType` throug the `IActivator` argument?

khellang replied:
I've pushed a commit to show what it would look like 😄

khellang replied:
But the "what are we even testing at this point?" question is becoming more and more relevant 😆

JamesNK committed:
Check type name for known value before getting type

JamesNK commented:
Sorry for the delay. I looked at the latest updates.

A problem I see is that custom activators won't have `ITypeForwardingActivator`. It will be a behavior change for them.

Another problem is type names get resolved over and over which is bad if `Type.GetType` throws an expensive exception. I want to match the old behavior without a regression in perf.

I made some changes that I think solves these two issues. Can you check out: https://github.com/dotnet/aspnetcore/compare/jamesnk/check-type-name-before-resolve

Changes:
* Before calling `Type.GetType`, the name is checked whether it is a known type. There is still a try/catch to handle an error, but this should reduce or eliminate exceptions from custom implementations.
* Added `ITypeNameResolver` interface for resolving type names. If the `IActivator` instance implements it then it then the passed in implementation is used, otherwise it falls back to a default implementation. Only used in tests.

What do you think? If you're happy with the changes I can start a new PR or you're welcome to update this one.

khellang committed:
Remove unnecessary suppression

khellang commented:
> What do you think? If you're happy with the changes I can start a new PR or you're welcome to update this one.

I'm most concerned about getting the bug patched 😄 I've rebased the branch, picked up your commit and removed a now-unnecessary suppression 👍

JamesNK commented:
@amcasey Can you take another look. This needs to be up to the standard of being patched in 8.0.

khellang commented:
> Do you need this patched in a 8.0 servicing release?

That would be awesome. Otherwise we'd have to wait until november? 😅

amcasey commented on src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.Tests/XmlEncryption/XmlEncryptionExtensionsTests.cs:
```diff
@@ -49,6 +49,99 @@ public void DecryptElement_RootNodeRequiresDecryption_Success()
+        mockActivator.ReturnDecryptedElementGivenDecryptorTypeNameAndInput(decryptorTypeName, "<node />", "<newNode />");
+        var mockTypeNameResolver = mockActivator.As<ITypeNameResolver>();
+
+        var serviceCollection = new ServiceCollection();
```
I must be missing something, but why is there a service collection?  Why not just use the activator directly?  Is something retrieving it from DI in the background?

JamesNK replied:
All the tests in this file do it. I went with the flow 😄

amcasey commented on src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.Tests/XmlEncryption/XmlEncryptionExtensionsTests.cs:
```diff
@@ -49,6 +49,99 @@ public void DecryptElement_RootNodeRequiresDecryption_Success()
+        mockActivator.Setup(o => o.CreateInstance(typeof(NullXmlDecryptor), decryptorTypeName)).Returns(new NullXmlDecryptor());
+        var mockTypeNameResolver = mockActivator.As<ITypeNameResolver>();
+        var resolvedType = typeof(NullXmlDecryptor);
+        mockTypeNameResolver.Setup(mockTypeNameResolver => mockTypeNameResolver.TryResolveType(It.IsAny<string>(), out resolvedType)).Returns(true);
```
Will the first argument be `typeof(NullXmlDecryptor).FullName`?  Why accept any string?

amcasey commented on src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.Tests/XmlEncryption/XmlEncryptionExtensionsTests.cs:
```diff
@@ -49,6 +49,99 @@ public void DecryptElement_RootNodeRequiresDecryption_Success()
+        var mockActivator = new Mock<IActivator>();
+        mockActivator.Setup(o => o.CreateInstance(typeof(IXmlDecryptor), decryptorTypeName)).Returns(new NullXmlDecryptor());
+        var mockTypeNameResolver = mockActivator.As<ITypeNameResolver>();
+        var resolvedType = typeof(NullXmlDecryptor);
```
Arguably, this should be null, though I don't think the contract requires it.

amcasey commented on src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.Tests/XmlEncryption/XmlEncryptionExtensionsTests.cs:
```diff
@@ -49,6 +49,99 @@ public void DecryptElement_RootNodeRequiresDecryption_Success()
+    public void DecryptElement_KnownType_UnableToResolveType_Success()
+    {
+        // Arrange
+        var decryptorTypeName = typeof(NullXmlDecryptor).AssemblyQualifiedName;
```
It looks like we could add coverage for the forwarding case by replacing "Microsoft.AspNetCore.DataProtection" with "Microsoft.AspNet.DataProtection" in `decryptorTypeName` (probably in a separate test).

JamesNK replied:
I updated a test to not accept any string. It now verifies that a forwarded name is used (assembly version is trimmed).

amcasey commented on src/DataProtection/DataProtection/src/KeyManagement/XmlKeyManager.cs:
```diff
@@ -463,27 +465,25 @@ IAuthenticatedEncryptorDescriptor IInternalXmlKeyManager.DeserializeDescriptorFr
-    [UnconditionalSuppressMessage("Trimmer", "IL2057", Justification = "Type.GetType result is only useful with types that are referenced by DataProtection assembly.")]
     private IAuthenticatedEncryptorDescriptorDeserializer CreateDeserializer(string descriptorDeserializerTypeName)
     {
         var resolvedTypeName = TypeForwardingActivator.TryForwardTypeName(descriptorDeserializerTypeName, out var forwardedTypeName)
```
And I think we pass the un-forwarded name to the activator in case it had special handling for the original name that would get skipped if we just instantiated the type we've already resolved?

amcasey reviewed (commented):
Generally looks good, but I had some questions.

JamesNK commented on src/DataProtection/DataProtection/src/TypeExtensions.cs:
```diff
@@ -39,4 +40,15 @@ public static Type GetTypeWithTrimFriendlyErrorMessage(string typeName)
+    public static bool MatchType(string resolvedTypeName, ITypeNameResolver typeNameResolver, Type matchType)
+    {
+        // Before attempting to resolve the name to a type, check if it starts with the full name of the type.
+        if (matchType.FullName != null && resolvedTypeName.StartsWith(matchType.FullName, StringComparison.Ordinal))
```
> Oh, maybe `resolvedTypeName` is expected to be assembly-qualified? (And we don't want compare it to `matchType.AssemblyQualifiedName` because of perf or because it might differ or something?)

Yes. I'll add a comment.

> Do we care about the case where the type is a closed generic, in which case the type arguments will be assembly-qualified?

I don't think any of the types we're comparing against has closed generics. I'll double check.

JamesNK commented on src/DataProtection/DataProtection/src/KeyManagement/XmlKeyManager.cs:
```diff
@@ -112,6 +113,7 @@ public XmlKeyManager(IOptions<KeyManagementOptions> keyManagementOptions, IActiv
         var escrowSinks = keyManagementOptions.Value.KeyEscrowSinks;
         _keyEscrowSink = escrowSinks.Count > 0 ? new AggregateKeyEscrowSink(escrowSinks) : null;
         _activator = activator;
+        _typeNameResolver = activator as ITypeNameResolver ?? DefaultTypeNameResolver.Instance;
```
> Or maybe this is purely here as a test hook (in which case, please add a comment)?

Just for tests.

JamesNK committed:
Comment

JamesNK commented on src/DataProtection/DataProtection/src/KeyManagement/XmlKeyManager.cs:
```diff
@@ -463,27 +465,25 @@ IAuthenticatedEncryptorDescriptor IInternalXmlKeyManager.DeserializeDescriptorFr
-    [UnconditionalSuppressMessage("Trimmer", "IL2057", Justification = "Type.GetType result is only useful with types that are referenced by DataProtection assembly.")]
     private IAuthenticatedEncryptorDescriptorDeserializer CreateDeserializer(string descriptorDeserializerTypeName)
     {
         var resolvedTypeName = TypeForwardingActivator.TryForwardTypeName(descriptorDeserializerTypeName, out var forwardedTypeName)
```
Changed it to typeNameToMatch.

JamesNK committed:
PR feedback

JamesNK commented:
@amcasey PR feedback applied.

amcasey commented on src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.Tests/XmlEncryption/XmlEncryptionExtensionsTests.cs:
```diff
@@ -107,7 +108,7 @@ public void DecryptElement_KnownType_TypeNameResolverCalled()
         // Assert
         XmlAssert.Equal("<value />", retVal);
-        mockTypeNameResolver.Verify(o => o.TryResolveType(It.IsAny<string>(), out resolvedType), Times.Once());
+        mockTypeNameResolver.Verify(o => o.TryResolveType(forwardedTypeName, out resolvedType), Times.Once());
```
I'd probably leave this one as any.

amcasey commented on src/DataProtection/DataProtection/src/TypeExtensions.cs:
```diff
@@ -41,9 +41,10 @@ public static Type GetTypeWithTrimFriendlyErrorMessage(string typeName)
     }
 
-    public static bool MatchType(string resolvedTypeName, ITypeNameResolver typeNameResolver, Type matchType)
+    public static bool MatchName(this Type matchType, string resolvedTypeName, ITypeNameResolver typeNameResolver)
```
"MatchesName"?

JamesNK commented on src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.Tests/XmlEncryption/XmlEncryptionExtensionsTests.cs:
```diff
@@ -49,6 +49,100 @@ public void DecryptElement_RootNodeRequiresDecryption_Success()
+
+        // Assert
+        XmlAssert.Equal("<value />", retVal);
+        mockTypeNameResolver.Verify(o => o.TryResolveType(forwardedTypeName, out resolvedType), Times.Once());
```
```suggestion
        mockTypeNameResolver.Verify(o => o.TryResolveType(It.IsAny<string>(), out resolvedType), Times.Once());
```

JamesNK committed:
Update src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.Tests/XmlEncryption/XmlEncryptionExtensionsTests.cs

JamesNK commented:
This missed the deadline for 8.0.4. We'll see about getting it into 8.0.5.

There is a long lead time for servicing.

JamesNK commented:
/backport to release/8.0
